src/ucp/core/ucp_request.inl

```diff
@@ -780,17 +795,57 @@ ucp_request_get_memory_type(ucp_context_h context, const void *address,
     return param->memory_type;
 }
 
+static UCS_F_ALWAYS_INLINE void ucp_request_id_alloc(ucp_request_t *req)
+{
+    ucp_worker_h worker = req->send.ep->worker;
+    ucs_status_t UCS_V_UNUSED status;
+
+    ucp_request_id_check(req, ==, UCP_REQUEST_ID_INVALID);
+    status = ucs_ptr_map_put(&worker->ptr_map, req,
+                             ucp_ep_use_indirect_id(req->send.ep),
+                             &req->id);
+    ucs_assertv(status == UCS_OK, "%p: failed to get id", req);
+}
+
 static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t
-ucp_send_request_get_id(ucp_request_t *req)
+ucp_request_get_id(const ucp_request_t *req)
 {
-    return ucp_worker_get_request_id(req->send.ep->worker, req,
-                                     ucp_ep_use_indirect_id(req->send.ep));
+    ucp_request_id_check(req, !=, UCP_REQUEST_ID_INVALID);
+    return req->id;
 }
 
 static UCS_F_ALWAYS_INLINE void
-ucp_send_request_set_id(ucp_request_t *req)
+ucp_request_id_release(ucp_request_t *req)
 {
-    req->send.msg_proto.sreq_id = ucp_send_request_get_id(req);
+    ucp_worker_h worker = req->send.ep->worker;
+    ucs_status_t UCS_V_UNUSED status;
+
+    ucp_request_id_check(req, !=, UCP_REQUEST_ID_INVALID);
+    status = ucs_ptr_map_del(&worker->ptr_map, req->id);
+    ucs_assertv(status == UCS_OK, "req %p: failed to release id", req);
+    ucp_request_id_reset(req);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_request_get_by_id(ucp_worker_h worker, ucs_ptr_map_key_t id,
+                      ucp_request_t **req_p, int extract)
+{
+    ucs_status_t status;
+    void *ptr;
+
+    status = ucs_ptr_map_get(&worker->ptr_map, id, extract, &ptr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    *req_p = (ucp_request_t*)ptr;
+    ucp_request_id_check(*req_p, ==, id);
+
+    if (extract) {
+        ucp_request_id_reset(*req_p);
+    }
```

Comment:
why needed here? only for assert?

Reply:
not only for assert, we will use this fact to remove check whether we need to removed from PTR map or not

Reply:
in case of error?
pls add comment

Reply:
> in case of error?

yes, ID will be put to PTR MAP in case of error to avoid leaks